File: Python/30.substring-with-concatenation-of-all-words.py
#
# @lc app=leetcode id=30 lang=python
#
# [30] Substring with Concatenation of All Words
#
# https://leetcode.com/problems/substring-with-concatenation-of-all-words/description/
#
# algorithms
# Hard (24.46%)
# Likes:    674
# Dislikes: 1065
# Total Accepted:    154.5K
# Total Submissions: 631.6K
# Testcase Example:  '"barfoothefoobarman"\n["foo","bar"]'
#
# You are given a string, s, and a list of words, words, that are all of the
# same length. Find all starting indices of substring(s) in s that is a
# concatenation of each word in words exactly once and without any intervening
# characters.
# 
# 
# 
# Example 1:
# 
# 
# Input:
# ⁠ s = "barfoothefoobarman",
# ⁠ words = ["foo","bar"]
# Output: [0,9]
# Explanation: Substrings starting at index 0 and 9 are "barfoo" and "foobar"
# respectively.
# The output order does not matter, returning [9,0] is fine too.
# 
# 
# Example 2:
# 
# 
# Input:
# ⁠ s = "wordgoodgoodgoodbestword",
# ⁠ words = ["word","good","best","word"]
# Output: []
# 
# 
#

# @lc code=start
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if not words:
            return [] 

        hash_dict = {}
        for word in words:
            hash_dict[word] = hash_dict.get(word, 0) + 1

        res = []
        num_words = len(words)
        len_word = len(words[0])
        for i in range(len(s)-num_words*len_word+1):
            temp_dict = {}
            num = 0
            while num<num_words:
                word = s[i+num*len_word: i+(num+1)*len_word]
                if word in hash_dict:
                    temp_dict[word] = temp_dict.get(word, 0) + 1
                    if temp_dict[word] > hash_dict[word]:
                        break
                else:
                    break
                num+=1

            if num == num_words:
                res.append(i)
        return res 
                
# @lc code=end

sol = Solution()
s = "barfoothefoobarman"
words = ["foo","bar"]
print(sol.findSubstring(s, words))

File: C++/30.substring-with-concatenation-of-all-words.cpp
/*
 * @lc app=leetcode id=30 lang=cpp
 *
 * [30] Substring with Concatenation of All Words
 */
#include<vector>
#include<string>
#include<unordered_map>
using namespace std;

// @lc code=start
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
         int n_w = words.size(), l_w = words[0].size();
        
        unordered_map<string, int> hmap;
        for (string& word: words) hmap[word] += 1;
        
        vector<int> ans;
        for (int i=0; i<s.size()-n_w*l_w+1; ++i) {
            unordered_map<string, int> tmp = hmap;
            
            int j = 0;
            for (; j<n_w; ++j) {
                string w = s.substr(i+j*l_w, l_w);
                if (tmp.count(w) && tmp[w]>0) tmp[w] -= 1;
                else break;
            }
            if (j==n_w) ans.push_back(i);
        }
        return ans;
    }
};
// @lc code=end

